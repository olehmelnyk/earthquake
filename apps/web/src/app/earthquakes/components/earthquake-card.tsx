import { FC } from 'react';
import { Card, CardContent, CardFooter, CardHeader, CardTitle } from '@earthquake/ui';
import { Earthquake } from '../../components/EarthquakeTable';
import { EarthquakeSeverityBadge } from './earthquake-severity-badge';
import { EarthquakeEditButton } from './earthquake-edit-button';
import { EarthquakeDeleteButton } from './earthquake-delete-button';
import dayjs from 'dayjs';
import { formatDistance } from 'date-fns';

export interface EarthquakeCardProps {
  readonly id: string;
  readonly location: string;
  readonly magnitude: number;
  readonly date: Date;
  readonly onDelete: (id: string) => Promise<void>;
  readonly onEdit: (id: string) => void;
}

export function EarthquakeCard({
  id,
  location,
  magnitude,
  date,
  onDelete,
  onEdit
}: EarthquakeCardProps) {
  return (
    <Card className="overflow-hidden">
      <CardHeader className="flex flex-row items-start justify-between space-y-0 pb-2">
        <div className="space-y-1">
          <CardTitle className="text-sm font-medium">
            {location}
          </CardTitle>
          <div className="flex items-center space-x-2">
            <EarthquakeSeverityBadge magnitude={magnitude} />
            <span className="text-xs text-muted-foreground">
              {formatDistance(date, new Date(), { addSuffix: true })}
            </span>
          </div>
        </div>
      </CardHeader>
      <CardContent>
        <div className="text-2xl font-bold" data-testid="magnitude">{magnitude.toFixed(1)}</div>
        <p className="text-xs text-muted-foreground mt-1">
          {date.toLocaleString()}
        </p>
      </CardContent>
      <CardFooter className="flex justify-between">
        <EarthquakeEditButton id={id} onEdit={onEdit} />
        <EarthquakeDeleteButton
          id={id}
          location={location}
          onDelete={onDelete}
        />
      </CardFooter>
    </Card>
  );
}