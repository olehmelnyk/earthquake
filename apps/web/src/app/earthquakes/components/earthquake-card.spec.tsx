import { render, screen } from '@testing-library/react';
import { EarthquakeCard } from './earthquake-card';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { userEvent } from '@testing-library/user-event';

// Mock child components
vi.mock('./earthquake-severity-badge', () => ({
  EarthquakeSeverityBadge: ({ magnitude }: { magnitude: number }) => {
    let severity = 'Micro';
    if (magnitude >= 7) severity = 'Major';
    else if (magnitude >= 5) severity = 'Moderate';
    else if (magnitude >= 3) severity = 'Minor';

    return <span data-testid="severity">{severity}</span>;
  }
}));

vi.mock('./earthquake-edit-button', () => ({
  EarthquakeEditButton: ({ id, onEdit }: { id: string; onEdit: (id: string) => void }) => (
    <button onClick={() => onEdit(id)}>Edit</button>
  )
}));

vi.mock('./earthquake-delete-button', () => ({
  EarthquakeDeleteButton: ({
    id,
    location,
    onDelete
  }: {
    id: string;
    location: string;
    onDelete: (id: string) => Promise<void>
  }) => {
    const handleClick = () => {
      // Simulate opening the dialog
      document.body.innerHTML += `
        <div>
          <div>Are you sure?</div>
          <div>permanently delete the earthquake record for <strong>${location}</strong></div>
        </div>
      `;
    };

    return <button onClick={handleClick}>Delete</button>;
  }
}));

// Mock UI components
vi.mock('@earthquake/ui', () => ({
  Card: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div className={`card ${className || ''}`}>{children}</div>
  ),
  CardHeader: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div className={`card-header ${className || ''}`}>{children}</div>
  ),
  CardContent: ({ children }: { children: React.ReactNode }) => (
    <div className="card-content">{children}</div>
  ),
  CardFooter: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div className={`card-footer ${className || ''}`}>{children}</div>
  ),
  CardTitle: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="location" className={`card-title ${className || ''}`}>{children}</div>
  ),
}));

// Mock date-fns
vi.mock('date-fns', () => ({
  formatDistance: () => 'about 1 hour ago',
}));

describe('EarthquakeCard', () => {
  const mockProps = {
    id: '1',
    location: '35.6762, 139.6503',
    magnitude: 6.5,
    date: new Date('2023-04-15T10:30:00'),
    onDelete: vi.fn().mockResolvedValue(undefined),
    onEdit: vi.fn(),
  };

  beforeEach(() => {
    vi.clearAllMocks();
    document.body.innerHTML = '';
  });

  it('should render earthquake information correctly', () => {
    render(<EarthquakeCard {...mockProps} />);

    expect(screen.getByTestId('location')).toHaveTextContent('35.6762, 139.6503');
    expect(screen.getByTestId('magnitude')).toHaveTextContent('6.5');
    expect(screen.getByTestId('severity')).toHaveTextContent('Moderate');
  });

  it('should call onEdit when edit button is clicked', async () => {
    const user = userEvent.setup();
    render(<EarthquakeCard {...mockProps} />);

    const editButton = screen.getByRole('button', { name: /edit/i });
    await user.click(editButton);

    expect(mockProps.onEdit).toHaveBeenCalledWith('1');
  });

  it('should display delete confirmation dialog when delete button is clicked', async () => {
    const user = userEvent.setup();
    render(<EarthquakeCard {...mockProps} />);

    const deleteButton = screen.getByRole('button', { name: /delete/i });
    await user.click(deleteButton);

    expect(screen.getByText('Are you sure?')).toBeInTheDocument();
    expect(screen.getByText(/permanently delete the earthquake record for/i)).toBeInTheDocument();
    expect(screen.getByText('35.6762, 139.6503', { selector: 'strong' })).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date('2023-04-15T10:30:00'));

    const { container } = render(<EarthquakeCard {...mockProps} />);
    expect(container).toMatchSnapshot();

    vi.useRealTimers();
  });
});