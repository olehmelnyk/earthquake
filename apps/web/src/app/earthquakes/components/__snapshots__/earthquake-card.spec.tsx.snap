// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EarthquakeCard > should match snapshot 1`] = `
<div>
  <div
    class="card overflow-hidden"
  >
    <div
      class="card-header flex flex-row items-start justify-between space-y-0 pb-2"
    >
      <div
        class="space-y-1"
      >
        <div
          class="card-title text-sm font-medium"
          data-testid="location"
        >
          35.6762, 139.6503
        </div>
        <div
          class="flex items-center space-x-2"
        >
          <span
            data-testid="severity"
          >
            Moderate
          </span>
          <span
            class="text-xs text-muted-foreground"
          >
            about 1 hour ago
          </span>
        </div>
      </div>
    </div>
    <div
      class="card-content"
    >
      <div
        class="text-2xl font-bold"
        data-testid="magnitude"
      >
        6.5
      </div>
      <p
        class="text-xs text-muted-foreground mt-1"
      >
        4/15/2023, 10:30:00 AM
      </p>
    </div>
    <div
      class="card-footer flex justify-between"
    >
      <button>
        Edit
      </button>
      <button>
        Delete
      </button>
    </div>
  </div>
</div>
`;
